### PR TITLE
Allow pdb delta to be optionally included in a hot reload payload

### LINKED ISSUE #39036: Pass PDB Deltas to MetadataUpdateHandler
After Hot Reload happens, breakpoints in the code which got modified are moved, but the execution breaks in the original location, which is incorrect.
This is due to not passing PDB deltas to the `MetadatUpdateHandler`. The hot reload agent UpdateDelta type and – ultimately – the websocket protocol would need to be updated to include the PDB delta as optional content.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/39036

### CODE REVIEW

**Comment by pranavkm** [comment]
FYI @BillHiebert

File: src/Components/Web.JS/src/Boot.WebAssembly.ts [hunk 1]
```diff
@@ -43,8 +43,8 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
     }
   });
 
-  Blazor._internal.applyHotReload = (id: string, metadataDelta: string, ilDeta: string) => {
-    DotNet.invokeMethod('Microsoft.AspNetCore.Components.WebAssembly', 'ApplyHotReloadDelta', id, metadataDelta, ilDeta);
+  Blazor._internal.applyHotReload = (id: string, metadataDelta: string, ilDeta: string, pdbDelta: string | undefined) => {
```
Comment (adrianwright109) [inline src/Components/Web.JS/src/Boot.WebAssembly.ts:46]: Typo, change `ilDeta` to `ilDelta`.

File: src/Components/Web.JS/src/GlobalExports.ts [hunk 2]
```diff
@@ -63,7 +63,7 @@ interface IBlazor {
     attachWebRendererInterop?: typeof attachWebRendererInterop,
 
     // APIs invoked by hot reload
-    applyHotReload?: (id: string, metadataDelta: string, ilDelta: string) => void,
+    applyHotReload?: (id: string, metadataDelta: string, ilDelta: string, pdbDelta: string | undefined) => void,
```
Comment (adrianwright109) [inline src/Components/Web.JS/src/GlobalExports.ts:66]: Typo, change `ilDeta` to `ilDelta`.

File: src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs [hunk 3]
```diff
@@ -42,13 +42,14 @@ internal static async Task InitializeAsync()
     /// For framework use only.
     /// </summary>
     [JSInvokable(nameof(ApplyHotReloadDelta))]
-    public static void ApplyHotReloadDelta(string moduleIdString, byte[] metadataDelta, byte[] ilDeta)
+    public static void ApplyHotReloadDelta(string moduleIdString, byte[] metadataDelta, byte[] ilDeta, byte[] pdbBytes)
```
Comment (adrianwright109) [inline src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs:45]: Typo, change `ilDeta` to `ilDelta`.

File: src/Components/Web.JS/src/Boot.WebAssembly.ts (hunk 1 above)
Comment (pranavkm) [reply src/Components/Web.JS/src/Boot.WebAssembly.ts:46]:
```suggestion
  Blazor._internal.applyHotReload = (id: string, metadataDelta: string, ilDelta: string, pdbDelta: string | undefined) => {
```

File: src/Components/Web.JS/src/Boot.WebAssembly.ts [hunk 4]
```diff
@@ -43,8 +43,8 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
...
+    DotNet.invokeMethod('Microsoft.AspNetCore.Components.WebAssembly', 'ApplyHotReloadDelta', id, metadataDelta, ilDeta, pdbDelta);
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Boot.WebAssembly.ts:47]:
```suggestion
    DotNet.invokeMethod('Microsoft.AspNetCore.Components.WebAssembly', 'ApplyHotReloadDelta', id, metadataDelta, ilDelta, pdbDelta);
```

File: src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs (hunk 3 above)
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs:45]:
```suggestion
    public static void ApplyHotReloadDelta(string moduleIdString, byte[] metadataDelta, byte[] ilDelta, byte[] pdbBytes)
```

File: src/Components/WebAssembly/WebAssembly/src/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -1 +1,3 @@
 #nullable enable
+*REMOVED*static Microsoft.AspNetCore.Components.WebAssembly.HotReload.WebAssemblyHotReload.ApplyHotReloadDelta(string! moduleIdString, byte[]! metadataDelta, byte[]! ilDeta) -> void
+static Microsoft.AspNetCore.Components.WebAssembly.HotReload.WebAssemblyHotReload.ApplyHotReloadDelta(string! moduleIdString, byte[]! metadataDelta, byte[]! ilDeta, byte[]! pdbBytes) -> void
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/PublicAPI.Unshipped.txt:3]:
```suggestion
static Microsoft.AspNetCore.Components.WebAssembly.HotReload.WebAssemblyHotReload.ApplyHotReloadDelta(string! moduleIdString, byte[]! metadataDelta, byte[]! ilDelta, byte[]! pdbBytes) -> void
```

File: src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs [hunk 6]
```diff
@@ -42,13 +42,14 @@ internal static async Task InitializeAsync()
...
     {
         var moduleId = Guid.Parse(moduleIdString);
 
         _updateDeltas[0].ModuleId = moduleId;
         _updateDeltas[0].MetadataDelta = metadataDelta;
         _updateDeltas[0].ILDelta = ilDeta;
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs:51]:
```suggestion
        _updateDeltas[0].ILDelta = ilDelta;
```

**Review by SteveSandersonMS — APPROVED** [review]

**Comment by pranavkm** [comment]
@BillHiebert the change should be available with a 7.0 SDK that becomes available later today / tomorrow. If you're able to confirm it works end-to-end, we can have this backported to 6.0.x at the earliest patch release.

**Pull request merged**